<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="An awesome OpenOmics chrom-seek pipeline for cfChIP-seq, ChIP-seq, and ATAC-seq data"><meta name=author content="Skyler Kuhn"><link rel=icon href=../assets/favicon/favicon.ico><meta name=generator content="mkdocs-1.4.3, mkdocs-material-8.3.0"><title>Best Practices - chrom-seek</title><link rel=stylesheet href=../assets/stylesheets/main.fd896c8a.min.css><link rel=stylesheet href=../assets/stylesheets/palette.cbb835fc.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../css/extra.css><script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#best-practices class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-component=outdated hidden> <aside class="md-banner md-banner--warning"> </aside> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=.. title=chrom-seek class="md-header__button md-logo" aria-label=chrom-seek data-md-component=logo> <img src=../assets/icons/doc-book.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> chrom-seek </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Best Practices </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 6a6 6 0 0 1 6 6c0 2.22-1.21 4.16-3 5.2V19a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1v-1.8c-1.79-1.04-3-2.98-3-5.2a6 6 0 0 1 6-6m2 15v1a1 1 0 0 1-1 1h-2a1 1 0 0 1-1-1v-1h4m6-10h3v2h-3v-2M1 11h3v2H1v-2M13 1v3h-2V1h2M4.92 3.5l2.13 2.14-1.42 1.41L3.5 4.93 4.92 3.5m12.03 2.13 2.12-2.13 1.43 1.43-2.13 2.12-1.42-1.42Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3 3.19.09m3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95 2.06.05m-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/OpenOmics/chrom-seek title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> OpenOmics/chrom-seek </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=.. class=md-tabs__link> About </a> </li> <li class=md-tabs__item> <a href=../usage/run/ class=md-tabs__link> Commands </a> </li> <li class=md-tabs__item> <a href=./ class="md-tabs__link md-tabs__link--active"> Best Practices </a> </li> <li class=md-tabs__item> <a href=../tools_software/ class=md-tabs__link> Tools/References </a> </li> <li class=md-tabs__item> <a href=../faq/glossary/ class=md-tabs__link> FAQ </a> </li> <li class=md-tabs__item> <a href=../license/ class=md-tabs__link> License </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=.. title=chrom-seek class="md-nav__button md-logo" aria-label=chrom-seek data-md-component=logo> <img src=../assets/icons/doc-book.svg alt=logo> </a> chrom-seek </label> <div class=md-nav__source> <a href=https://github.com/OpenOmics/chrom-seek title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> OpenOmics/chrom-seek </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=.. class=md-nav__link> About </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_2 type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2> Commands <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Commands data-md-level=1> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Commands </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../usage/run/ class=md-nav__link> chrom-seek run </a> </li> <li class=md-nav__item> <a href=../usage/unlock/ class=md-nav__link> chrom-seek unlock </a> </li> <li class=md-nav__item> <a href=../usage/install/ class=md-nav__link> chrom-seek install </a> </li> <li class=md-nav__item> <a href=../usage/cache/ class=md-nav__link> chrom-seek cache </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Best Practices <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> Best Practices </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#our-take-home-message class=md-nav__link> Our take-home message </a> </li> <li class=md-nav__item> <a href=#key-points-to-consider class=md-nav__link> Key points to consider </a> </li> <li class=md-nav__item> <a href=#recommendations class=md-nav__link> Recommendations </a> <nav class=md-nav aria-label=Recommendations> <ul class=md-nav__list> <li class=md-nav__item> <a href=#general-considerations class=md-nav__link> General considerations </a> </li> <li class=md-nav__item> <a href=#peak-calling class=md-nav__link> Peak calling </a> </li> <li class=md-nav__item> <a href=#differential-analyses class=md-nav__link> Differential analyses </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#useful-links class=md-nav__link> Useful links </a> <nav class=md-nav aria-label="Useful links"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#encode-standards-and-recommendations class=md-nav__link> ENCODE standards and recommendations </a> </li> <li class=md-nav__item> <a href=#nf-core-output-file-list-with-useful-documentation-about-qc-metrics-we-use-in-chrom-seek class=md-nav__link> NF-core output file list with useful documentation about QC metrics we use in chrom-seek </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../tools_software/ class=md-nav__link> Tools/References </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5 type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5> FAQ <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=FAQ data-md-level=1> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../faq/glossary/ class=md-nav__link> Glossary of terms </a> </li> <li class=md-nav__item> <a href=../Useful_review_articles/ class=md-nav__link> Useful reviews </a> </li> <li class=md-nav__item> <a href=../faq/Input_vs_IgG/ class=md-nav__link> Input vs IgG </a> </li> <li class=md-nav__item> <a href=../manual_workflows/ class=md-nav__link> Manually running DiffBind </a> </li> <li class=md-nav__item> <a href=../faq/questions/ class=md-nav__link> General Questions </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../license/ class=md-nav__link> License </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/OpenOmics/chrom-seek/edit/main/docs/best_practices.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1 id=best-practices>Best practices<a class=headerlink href=#best-practices title="Permanent link">&para;</a></h1> <h2 id=our-take-home-message>Our take-home message<a class=headerlink href=#our-take-home-message title="Permanent link">&para;</a></h2> <ol> <li>Always remember to consider the specific immunoprecipation or methodology for each analysis.</li> <li>Bigwigs serve as our closest approximation to truth sets. It's essential to cross-check results against them.</li> <li>The boundaries of peaks are primarily influenced by the tool used for analysis rather than being inherent biological features.</li> <li>This type of data often contains significant noise and a high background level.</li> <li>The majority of reads are not detected within peaks. Among the remaining reads, some may represent background artifacts, while others have biological significance. However, distinguishing between the two is challenging. Additionally, the proportion of reads within peaks is influenced by biological factors.</li> </ol> <h2 id=key-points-to-consider>Key points to consider<a class=headerlink href=#key-points-to-consider title="Permanent link">&para;</a></h2> <p><strong>Not all ChIP-seq data can be treated equally!</strong></p> <p><em>Here are some main classes of proteins to consider when analyzing ChIP-seq data:</em> </p> <ul> <li>Transcription factors</li> <li>Promoter-associated histones </li> <li>Enhancer-associated histones </li> <li>Repressive histones </li> <li>RNA Polymerase II </li> <li>Repressive histone marks </li> <li>Chromosome structure proteins </li> <li>DNA damage proteins </li> </ul> <p>Transcription factors are often regarded as the most straightforward targets in ChIP-seq experiments due to their high abundance, strong binding affinity, and well-characterized binding motifs. They typically exhibit clear and distinct peaks in ChIP-seq data, facilitating their detection and analysis. Moreover, transcription factors are known for their resilience to variations in processing and analysis protocols compared to other genomic features. They also demonstrate reduced dependence on ideal experimental controls. Additionally, transcription factors are frequently associated with established gold standards, which aid in sensitivity and specificity assessments in ChIP-seq assays. Their specific binding to DNA sequences further simplifies validation in laboratory experiments.</p> <p>Databases like <a href=https://www.encodeproject.org/ >ENCODE</a> and <a href=http://cistrome.org/ >Cistrome</a> are valuable resources for studying specific marks. These databases employ robust data processing methods, although the sensitivity and specificity of processed features may be relatively lower compared to other methods. Some features may be poorly identified, particularly in a subset of cases.</p> <p>High noise and background levels affect all these methods, influenced by both biological factors and sample processing. Considering this influence is crucial at every stage of experimental design and data processing.</p> <p>ATAC-seq presents two primary challenges. Firstly, it explores two types of open chromatin regions, each characterized by distinct patterns and functions. These regions encompass histones flanked by spaces, typically found within active genes, and larger open regions devoid of stable histones, such as promoters. Secondly, ATAC-seq lacks a comprehensive global truth set for validation. Consequently, there is a scarcity of literature on best practices for ATAC-seq, with methodologies often relying on knowledge from similar methods.</p> <p><em>When annotating peaks, it's essential to remember:</em> </p> <ul> <li>A single peak may influence multiple genes.</li> <li>Unless it occurs within a gene's promoter or within 1kb of a transcriptional start site, determining if the nearest gene is actually affected by the peak is challenging without additional information.</li> <li>Activating transcription factors within gene bodies are unlikely to affect that gene directly.</li> <li>Peak boundaries are arbitrary.</li> <li>While summits are likely binding sites for transcription factors, they may not hold true for other features.</li> </ul> <h2 id=recommendations>Recommendations<a class=headerlink href=#recommendations title="Permanent link">&para;</a></h2> <h3 id=general-considerations>General considerations<a class=headerlink href=#general-considerations title="Permanent link">&para;</a></h3> <p>Using QC metrics as the sole criterion for sample quality assessment is unreliable. While QC values may be informative for certain features or laboratories, they do not universally apply. While they provide a starting point, they should <strong>not</strong> be used as an automatic decision-making tool for sample exclusion!</p> <p>Normalized counts matrices and derived results should not be solely relied upon, as they depend on the quality of peak calls, consensus peak identification, and normalization methods used. It's crucial to confirm the validity of results using normalized bigwigs or directly generated plots. </p> <p><em>Bigwigs offer enhanced visibility as they have been processed relative to the bam files in several ways:</em> </p> <ul> <li>Normalization to library size, utilizing only useful reads.</li> <li>Position adjustment as needed, particularly significant for single-end ChIP.</li> <li>Normalization against input to mitigate one form of biological noise, if available.</li> <li>75 bp smoothing, aligning with the resolution of most methods (50-250bp).</li> </ul> <h3 id=peak-calling>Peak calling<a class=headerlink href=#peak-calling title="Permanent link">&para;</a></h3> <p>Here's a structured approach for choosing a peak caller that is most suitable for your data. Remember to consider the specific immunoprecipation or methodology for each analysis. The following guidelines are based on the most common types of data, but they may not be applicable to all scenarios. Always consult the literature, consider the specific characteristics of your data, and visually inspect the peak calls along with their relevent tracks before making a descision.</p> <ol> <li><em>Transcription Factor (TF) Analysis</em><ul> <li>Inspect the macsNarrow calls in IGV.<ul> <li>If macsNarrow calls appear satisfactory, use macsNarrow.</li> <li>If not, consider using macsBroad.</li> </ul> </li> </ul> </li> <li><em>ATAC-seq Analysis</em><ul> <li>Determine if publication status matters.<ul> <li>If publication status matters, utilize macsNarrow.</li> <li>If not, compare results from macsNarrow and Genrich to determine the better option.</li> </ul> </li> </ul> </li> <li><em>CUN &amp; RUN or CUT &amp; TAG Analysis</em><ul> <li>Compare results from macsNarrow and SEACR to determine the better option.</li> </ul> </li> <li><em>Narrow Histone Mark Analysis</em><ul> <li>Uncertain about whether the data represents a narrow histone mark? Consult the <a href=https://www.encodeproject.org/chip-seq/histone/ >ENCODE documentation</a>.</li> <li>Compare results from macsNarrow and macsBroad.<ul> <li>If one outperforms the other consistently, choose the better option.</li> <li>If not, either option is viable.</li> </ul> </li> </ul> </li> <li><em>Other Marks Analysis</em><ul> <li>Check the macsBroad calls. <ul> <li>If they appear suitable, use macsBroad.</li> </ul> </li> <li>If macsBroad calls don't meet expectations, assess the pipeline default sicer calls in the chrome-seek pipeline.<ul> <li>If sicer calls are acceptable, utilize those.</li> <li>If not, adjust sicer conditions as necessary, noting that peak calling may not be feasible for all marks.</li> </ul> </li> </ul> </li> </ol> <h3 id=differential-analyses>Differential analyses<a class=headerlink href=#differential-analyses title="Permanent link">&para;</a></h3> <p>Establishing a robust method for identifying consensus peaks is essential. To effectively utilize any differential analysis tool, it's imperative to have a shared set of peaks across all samples. This step facilitates the removal of peaks with low counts, so avoid filtering outside of this process. Keep in mind that peak calling introduces noise, and peak boundaries lack biological significance. Additionally, be aware that peaks in one sample may overlap with multiple peaks in another sample.</p> <p>Normalization plays a crucial role in differential peak calling. Numerous studies indicate that both the consensus peak list and normalization method significantly impact results. However, determining the optimal approach can be challenging. A consistent finding in the literature is that normalizing solely to reads within peaks, without external information, consistently yields subpar results. This is likely because a significant proportion of reads (60-99%) often reside outside peaks, representing biologically relevant but rarer binding sites. I prefer scaling to the number of reads aligning to the genome following deduplication. In DiffBind v2, the TMM output takes the raw counts, corrects for inputs when applicable, and employs a DESeq2 function to scale the samples to an external value (the number of reads in the bam file) rather than relying solely on the internal values in the counts table.</p> <p>In ChIP-seq and ATAC-seq analyses, simpler models tend to yield better results. This is likely due to the various sources of noise in the data, including biological, technical, and computational factors. It's essential to scrutinize results from any method used to ensure that background noise isn't being erroneously identified.</p> <p>In the majority of projects, skews in global fold-changes towards one condition or the other are likely to be biologically meaningful. It's essential not to artificially equalize the number of peaks on both sides of the contrast unless there's prior knowledge or a substantial difference in the size of the original libraries (exceeding 10X).</p> <p>It's not uncommon to observe small differential fold-changes, even among the most significant peaks. Many papers assessing differential peak calls, as well as the default conditions in tools like DiffBind, often do not include fold-change filtering. However, in my experience, it can be advantageous to incorporate a low fold-change cutoff. A reasonable starting point is an absolute fold-change cutoff of 1.5 (equivalent to a log2 value of 0.585). Alternatively, using a cutoff of 4 (log2 value of 2) will typically remove almost all peaks unless the analysis is focused on transcription factors.</p> <p><strong>Please note:</strong> The pipeline opts to use a depreciated version of DiffBind for several reasons. This is by design!</p> <p><em>Here is why:</em></p> <ol> <li>DiffBind v2 has consistently demonstrated superior performance compared to other differential peak callers across a wide range of features documented in the literature.</li> <li>DiffBind v3 lacks rigorous testing, and the author does not provide rationale for changes in default counting or normalization parameters. It is feasible to achieve results similar to DiffBind v2, especially for ATAC data or in cases lacking input samples, by configuring the pipeline correctly. However, it consistently treats input samples differently regardless of configuration.</li> <li>The tool automatically generates a set of consensus peaks by combining peaks identified in at least two samples, regardless of condition, allowing for partial overlap. This results in each contrast having a unique set of comparison peaks, but it enhances the accuracy of differential peak calling. However, adjustments may be necessary for large cohorts with numerous samples.</li> <li>In DiffBind v2, the default normalization method scales the entire library to all reads in the alignment files. This approach is generally considered one of the most effective normalization options available.</li> </ol> <h2 id=useful-links>Useful links<a class=headerlink href=#useful-links title="Permanent link">&para;</a></h2> <p>Here are a set of useful links for learning more about ChIP-seq and ATAC-seq.</p> <h3 id=encode-standards-and-recommendations>ENCODE standards and recommendations<a class=headerlink href=#encode-standards-and-recommendations title="Permanent link">&para;</a></h3> <ul> <li><a href=https://www.encodeproject.org/chip-seq/transcription_factor/ >Transcription factors</a></li> <li><a href=https://www.encodeproject.org/chip-seq/histone/ >Histone marks</a></li> <li><a href=https://www.encodeproject.org/atac-seq/ >ATAC-seq</a></li> </ul> <h3 id=nf-core-output-file-list-with-useful-documentation-about-qc-metrics-we-use-in-chrom-seek>NF-core output file list with useful documentation about QC metrics we use in chrom-seek<a class=headerlink href=#nf-core-output-file-list-with-useful-documentation-about-qc-metrics-we-use-in-chrom-seek title="Permanent link">&para;</a></h3> <ul> <li><a href=https://nf-co.re/cutandrun/3.2.2/docs/output/ >Cut &amp; run</a></li> <li><a href=https://nf-co.re/chipseq/2.1.0/docs/output/ >ChIP-seq</a></li> <li><a href=https://nf-co.re/atacseq/2.1.2/docs/output/ >ATAC-seq</a></li> </ul> <hr> <div class=md-source-file> <small> Last update: 2026-02-05 </small> </div> </article> </div> </div> <a href=# class="md-top md-top--hidden md-icon" data-md-component=top> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../usage/cache/ class="md-footer__link md-footer__link--prev" aria-label="Previous: chrom-seek cache" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> chrom-seek cache </div> </div> </a> <a href=../tools_software/ class="md-footer__link md-footer__link--next" aria-label="Next: Tools/References" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Tools/References </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2024 OpenOmics </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://idss-bioinformatics.nih.gov/ target=_blank rel=noopener title=idss-bioinformatics.nih.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M319.9 320c57.41 0 103.1-46.56 103.1-104s-46.54-104-103.1-104c-57.41 0-103.1 46.56-103.1 104-.9 57.4 45.7 104 103.1 104zm50 32h-99.8C191.6 352 128 411.7 128 485.3c0 14.8 12.7 26.7 28.4 26.7h327.2c15.7 0 28.4-11.9 28.4-26.7 0-73.6-63.6-133.3-142.1-133.3zM512 160c44.18 0 80-35.82 80-80S556.2 0 512 0c-44.18 0-80 35.82-80 80s35.8 80 80 80zm-328.1 56c0-5.449.982-10.63 1.609-15.91C174.6 194.1 162.6 192 149.9 192H88.08C39.44 192 0 233.8 0 285.3 0 295.6 7.887 304 17.62 304h199.5c-20.42-23.8-33.22-54.3-33.22-88zM128 160c44.18 0 80-35.82 80-80S172.2 0 128 0C83.82 0 48 35.82 48 80s35.82 80 80 80zm423.9 32h-61.84c-12.8 0-24.88 3.037-35.86 8.24.6 5.26 1.6 10.36 1.6 15.76 0 33.71-12.78 64.21-33.16 88h199.7c9.76 0 17.66-8.4 17.66-18.7 0-51.5-39.4-93.3-88.1-93.3z"/></svg> </a> <a href=https://github.com/OpenOmics target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://hub.docker.com/u/skchronicles target=_blank rel=noopener title=hub.docker.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1v59.4zm0-204.3h-66.1v60.7h66.1V32zm78.2 144.8H362v59.4h66.1v-59.4zm-156.3-72.1h-66.1v60.1h66.1v-60.1zm78.1 0h-66.1v60.1h66.1v-60.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1l-13.3-8.9zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm-78.1-72.1h-66.1v60.1h66.1v-60.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["navigation.tabs", "navigation.top", "toc.integrate"], "search": "../assets/javascripts/workers/search.b028fd86.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../assets/javascripts/bundle.0f659f14.min.js></script> </body> </html>