<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="An awesome OpenOmics chrom-seek pipeline for cfChIP-seq, ChIP-seq, and ATAC-seq data"><meta name=author content="Skyler Kuhn"><link rel=icon href=../assets/favicon/favicon.ico><meta name=generator content="mkdocs-1.4.3, mkdocs-material-8.3.0"><title>Tools/References - chrom-seek</title><link rel=stylesheet href=../assets/stylesheets/main.fd896c8a.min.css><link rel=stylesheet href=../assets/stylesheets/palette.cbb835fc.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../css/extra.css><script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#tools-and-versions class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-component=outdated hidden> <aside class="md-banner md-banner--warning"> </aside> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=.. title=chrom-seek class="md-header__button md-logo" aria-label=chrom-seek data-md-component=logo> <img src=../assets/icons/doc-book.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> chrom-seek </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Tools/References </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 6a6 6 0 0 1 6 6c0 2.22-1.21 4.16-3 5.2V19a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1v-1.8c-1.79-1.04-3-2.98-3-5.2a6 6 0 0 1 6-6m2 15v1a1 1 0 0 1-1 1h-2a1 1 0 0 1-1-1v-1h4m6-10h3v2h-3v-2M1 11h3v2H1v-2M13 1v3h-2V1h2M4.92 3.5l2.13 2.14-1.42 1.41L3.5 4.93 4.92 3.5m12.03 2.13 2.12-2.13 1.43 1.43-2.13 2.12-1.42-1.42Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3 3.19.09m3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95 2.06.05m-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/OpenOmics/chrom-seek title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> OpenOmics/chrom-seek </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=.. class=md-tabs__link> About </a> </li> <li class=md-tabs__item> <a href=../usage/run/ class=md-tabs__link> Commands </a> </li> <li class=md-tabs__item> <a href=../best_practices/ class=md-tabs__link> Best Practices </a> </li> <li class=md-tabs__item> <a href=./ class="md-tabs__link md-tabs__link--active"> Tools/References </a> </li> <li class=md-tabs__item> <a href=../faq/glossary/ class=md-tabs__link> FAQ </a> </li> <li class=md-tabs__item> <a href=../license/ class=md-tabs__link> License </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=.. title=chrom-seek class="md-nav__button md-logo" aria-label=chrom-seek data-md-component=logo> <img src=../assets/icons/doc-book.svg alt=logo> </a> chrom-seek </label> <div class=md-nav__source> <a href=https://github.com/OpenOmics/chrom-seek title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> OpenOmics/chrom-seek </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=.. class=md-nav__link> About </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_2 type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2> Commands <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Commands data-md-level=1> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Commands </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../usage/run/ class=md-nav__link> chrom-seek run </a> </li> <li class=md-nav__item> <a href=../usage/unlock/ class=md-nav__link> chrom-seek unlock </a> </li> <li class=md-nav__item> <a href=../usage/install/ class=md-nav__link> chrom-seek install </a> </li> <li class=md-nav__item> <a href=../usage/cache/ class=md-nav__link> chrom-seek cache </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../best_practices/ class=md-nav__link> Best Practices </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Tools/References <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> Tools/References </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#quality-control-assessment-tools class=md-nav__link> Quality-control assessment tools </a> </li> <li class=md-nav__item> <a href=#data-processing-tools class=md-nav__link> Data processing tools </a> </li> <li class=md-nav__item> <a href=#peak-calling-and-differential-peak-calling-tools class=md-nav__link> Peak calling and differential peak calling tools </a> </li> <li class=md-nav__item> <a href=#annotations-motifs-and-qc-metrics class=md-nav__link> Annotations, motifs, and QC metrics </a> </li> <li class=md-nav__item> <a href=#reference-genome-resources class=md-nav__link> Reference genome resources </a> </li> <li class=md-nav__item> <a href=#main-differences-between-assays class=md-nav__link> Main differences between assays </a> </li> <li class=md-nav__item> <a href=#acknowledgements class=md-nav__link> Acknowledgements </a> <nav class=md-nav aria-label=Acknowledgements> <ul class=md-nav__list> <li class=md-nav__item> <a href=#biowulf class=md-nav__link> Biowulf </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5 type=checkbox id=__nav_5> <label class=md-nav__link for=__nav_5> FAQ <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=FAQ data-md-level=1> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../faq/glossary/ class=md-nav__link> Glossary of terms </a> </li> <li class=md-nav__item> <a href=../Useful_review_articles/ class=md-nav__link> Useful reviews </a> </li> <li class=md-nav__item> <a href=../faq/Input_vs_IgG/ class=md-nav__link> Input vs IgG </a> </li> <li class=md-nav__item> <a href=../manual_workflows/ class=md-nav__link> Manually running DiffBind </a> </li> <li class=md-nav__item> <a href=../faq/questions/ class=md-nav__link> General Questions </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../license/ class=md-nav__link> License </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/OpenOmics/chrom-seek/edit/main/docs/tools_software.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1 id=tools-and-versions>Tools and versions<a class=headerlink href=#tools-and-versions title="Permanent link">&para;</a></h1> <h2 id=quality-control-assessment-tools>Quality-control assessment tools<a class=headerlink href=#quality-control-assessment-tools title="Permanent link">&para;</a></h2> <p>The chrom-seek pipeline runs a series of quality-control (QC) tools to assess the sequencing and library quality of each sample. The tools used in the pipeline are listed below, along with their versions and a brief description of their purpose.</p> <table> <thead> <tr> <th><strong>Tool</strong></th> <th align=center><strong>Version</strong></th> <th><strong>Notes</strong></th> </tr> </thead> <tbody> <tr> <td>FastQC<sup>1<sup></td> <td align=center>0.11.9</td> <td>Assess sequencing quality, run before and after adapter trimming</td> </tr> <tr> <td>Kraken<sup>2<sup></td> <td align=center>2.1.2</td> <td>Assess microbial taxonomic composition</td> </tr> <tr> <td>KronaTools<sup>3<sup></td> <td align=center>2.8.1</td> <td>Visualize kraken output</td> </tr> <tr> <td>FastQ Screen<sup>4<sup></td> <td align=center>0.9.3</td> <td>Assess contamination; additional dependencies: <code>bowtie2/2.5.1</code>, <code>perl/5.36</code></td> </tr> <tr> <td>Preseq<sup>5<sup></td> <td align=center>3.1.2</td> <td>Estimate library complexity</td> </tr> <tr> <td>MultiQC<sup>6<sup></td> <td align=center>1.14</td> <td>Aggregate sample statistics and quality-control information across all samples</td> </tr> </tbody> </table> <p><strong>References</strong></p> <p><sup>1. <strong>FastQC:</strong> Andrews, S. (2010). FastQC: a quality control tool for high throughput sequence data.<a href=https://www.bioinformatics.babraham.ac.uk/projects/fastqc>https://www.bioinformatics.babraham.ac.uk/projects/fastqc</a></sup><br> <sup>2. <strong>Kraken:</strong> Wood, D. E. and S. L. Salzberg (2014). "Kraken: ultrafast metagenomic sequence classification using exact alignments." Genome Biol 15(3): R46. <a href=http://ccb.jhu.edu/software/kraken/ >http://ccb.jhu.edu/software/kraken/</a></sup><br> <sup>3. <strong>Krona:</strong> Ondov, B. D., et al. (2011). "Interactive metagenomic visualization in a Web browser." BMC Bioinformatics 12(1): 385. <a href=https://github.com/marbl/Krona/wiki>https://github.com/marbl/Krona/wiki</a></sup><br> <sup>4. <strong>FastQ Screen:</strong> Wingett, S. and S. Andrews (2018). "FastQ Screen: A tool for multi-genome mapping and quality control." F1000Research 7(2): 1338. <a href=https://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/ >https://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/</a></sup><br> <sup>5. <strong>Preseq:</strong> Daley, T. and A.D. Smith (2013). Predicting the molecular complexity of sequencing libraries. Nat Methods 10(4): 325-7. <a href=http://smithlabresearch.org/software/preseq/ >http://smithlabresearch.org/software/preseq/</a></sup><br> <sup>6. <strong>MultiQC:</strong> Ewels, P., et al. (2016). "MultiQC: summarize analysis results for multiple tools and samples in a single report." Bioinformatics 32(19): 3047-3048. <a href=https://multiqc.info/docs/ >https://multiqc.info/docs/</a></sup> </p> <h2 id=data-processing-tools>Data processing tools<a class=headerlink href=#data-processing-tools title="Permanent link">&para;</a></h2> <p>The pipeline is composed of a series of data processing steps that include adapter trimming, read alignment, duplicate removal, and bigwig creation. The tools used in the pipeline are listed below, along with their versions and a brief description of their purpose.</p> <table> <thead> <tr> <th><strong>Tool</strong></th> <th align=center><strong>Version</strong></th> <th align=left><strong>Notes</strong></th> </tr> </thead> <tbody> <tr> <td>Cutadapt<sup>1</sup></td> <td align=center>4.4</td> <td align=left>Remove adapter sequences and perform quality trimming</td> </tr> <tr> <td>BWA mem<sup>2</sup></td> <td align=center>0.7.17</td> <td align=left>Read alignment, first to identify reads aligning to blacklisted regions and later for the remainder of the genome</td> </tr> <tr> <td>Picard<sup>3</sup></td> <td align=center>2.27.3</td> <td align=left>Run SamToFastq (for blacklist read removal) and MarkDuplicates (to remove PCR duplicates in PE data)</td> </tr> <tr> <td>SAMtools<sup>4</sup></td> <td align=center>1.17</td> <td align=left>Remove reads with mapQ less than 6. Also run flagstat and idxstats to calculate alignment statistics.</td> </tr> <tr> <td>MACS<sup>5</sup></td> <td align=center>2.2.7.1</td> <td align=left>Run filterdup on SE data (<code>--keep-dup="auto"</code>) to remove PCR duplicates</td> </tr> <tr> <td>Bedtools<sup>6</sup></td> <td align=center>2.27.1</td> <td align=left>Run intersect and bedtobam to convert <code>.tag.Align.gz</code> to <code>.bam</code> for use with deeptools (specific to SE data) and MEME</td> </tr> <tr> <td>ppqt<sup>7,8</sup></td> <td align=center>1.1.2</td> <td align=left>Also known as phantompeakqualtools, used to calculate estimated fragment length (used for bigwig and peak calling for SE data). Also produces QC metrics: NSC and RSC.</td> </tr> <tr> <td>deepTools<sup>9</sup></td> <td align=center>3.5.1</td> <td align=left>Used for bigwig creation and multiple QC metrics. Use bamcoverage to create RPGC-normalized data: <code>--binSize 25 --smoothLength 75 --normalizeUsing RPGC</code>. For PE data, add <code>--centerReads</code>. For control SE, add <code>-e 200</code>. For ChIP SE, add <code>-e [estimated fragmentlength]</code>. For control subtraction (inputnorm), use bigwigCompare: <code>--binSize 25 --operation 'subtract'</code>. Run multiBigWigSummary, plotCorrelation, plotPCA, plotFingerprint, computeMatrix, plotHeatmap, and plotProfile for QC plots. Note: not all these have been reincorporated into the pipeline.</td> </tr> </tbody> </table> <p><strong>References</strong></p> <p><sup>1. <strong>Cutadapt:</strong> Martin, M. (2011). "Cutadapt removes adapter sequences from high-throughput sequencing reads." EMBnet 17(1): 10-12. <a href=https://cutadapt.readthedocs.io/en/stable/ >https://cutadapt.readthedocs.io/en/stable/</a> </sup><br> <sup>2. <strong>BWA:</strong> Li H. and Durbin R. (2009) Fast and accurate short read alignment with Burrows-Wheeler Transform. Bioinformatics 25: 1754-60. <a href=http://bio-bwa.sourceforge.net/bwa.shtml>http://bio-bwa.sourceforge.net/bwa.shtml</a> </sup><br> <sup>3. <strong>Picard:</strong> The Picard toolkit. <a href=https://broadinstitute.github.io/picard/ >https://broadinstitute.github.io/picard/</a></sup><br> <sup>4. <strong>SAMtools:</strong> Danecek, Petr et al. “Twelve years of SAMtools and BCFtools.” GigaScience vol. 10,2 (2021): giab008. </sup><br> <sup>5. <strong>MACS:</strong> Zhang, Y., et al. (2008). Model-based Analysis of ChIP-Seq (MACS). Genome Biol 9: R137. <a href=https://github.com/macs3-project/MACS>https://github.com/macs3-project/MACS</a></sup><br> <sup>6. <strong>Bedtools:</strong> Quinlan, A.R. (2014). BEDTools: The Swiss‐Army Tool for Genome Feature Analysis. Current Protocols in Bioinformatics, 47: 11.12.1-11.12.34. <a href=https://bedtools.readthedocs.io/en/latest/index.html>https://bedtools.readthedocs.io/en/latest/index.html</a></sup><br> <sup>7. <strong>Ppqt:</strong> Landt S.G., et al. (2012). ChIP-seq guidelines and practices of the ENCODE and modENCODE consortia. Genome Res 22(9): 1813-31. <a href=https://github.com/kundajelab/phantompeakqualtools>https://github.com/kundajelab/phantompeakqualtools</a></sup><br> <sup>8. <strong>Ppqt:</strong> Kharchenko P.K., et al. (2008). Design and analysis of ChIP-seq experiments for DNA-binding proteins Nat Biotechnol 26(12): 1351-9.</sup><br> <sup>9. <strong>deepTools:</strong> Ramírez, F., et al. (2016). deepTools2: A next Generation Web Server for Deep-Sequencing Data Analysis. Nucleic Acids Research 44(W1): W160--W165. <a href=https://deeptools.readthedocs.io/en/develop/ >https://deeptools.readthedocs.io/en/develop/</a></sup> </p> <h2 id=peak-calling-and-differential-peak-calling-tools>Peak calling and differential peak calling tools<a class=headerlink href=#peak-calling-and-differential-peak-calling-tools title="Permanent link">&para;</a></h2> <p>The pipeline includes peak calling and differential peak calling tools to identify enriched regions of interest in the genome. The tools used in the pipeline are listed below, along with their versions and a brief description of their purpose.</p> <table> <thead> <tr> <th><strong>Tool</strong></th> <th align=center><strong>Version</strong></th> <th align=left><strong>Notes</strong></th> </tr> </thead> <tbody> <tr> <td>MACS<sup>1</sup></td> <td align=center>2.2.7.1</td> <td align=left><strong><code>macsNarrow</code></strong>: The macs2 caller optimized for narrow peaks, widely recognized as the most popular peak calling algorithm. Typically used in large databases, it identifies peaks within the range of 150bp to 10kb. Originally designed to handle peaks with a single maxima/summit, its false discovery rate (FDR) has been greatly improved with the addition of an "input" control. It is generally more accurate than most other peak callers, even without controls. <strong><code>macsBroad</code></strong>: The macs2 caller for slightly broader peaks, sharing a similar algorithm with macsNarrow. It is particularly useful when peaks exhibit more than one maxima/summit.</td> </tr> <tr> <td>Sicer<sup>2</sup></td> <td align=center>2-1.0.3</td> <td align=left>Sicer is a broad peak caller that can be highly effective for certain histone marks. However, it may not perform well for extra broad domains such as lamins or some repressive marks. It allows for a small amount of gaps between peaks, and users may need to adjust window and gap parameters for optimal results.</td> </tr> <tr> <td>Genrich<sup>3</sup></td> <td align=center>0.6</td> <td align=left>Designed with ATAC-seq data in mind, Genrich can yield excellent results. However, it may not be universally favored by all collaborators due to its lack of formal publication or review.</td> </tr> <tr> <td>SEACR<sup>4</sup></td> <td align=center>1.3</td> <td align=left>Peak caller for CUT &amp; RUN or CUT &amp; TAG only; can utilize IgG as if it was an input control, but assumes IgG is sparse.</td> </tr> <tr> <td>MANorm<sup>5</sup></td> <td align=center>1.1.4</td> <td align=left>Differential peak calling when no replicates. This tool has not been incorporated into the pipeline.</td> </tr> <tr> <td>DiffBind<sup>6,7</sup></td> <td align=center>2.15.2</td> <td align=left>Used for conducting differential peak calling analyses, this tool integrates with Deseq2 and EdgeR for analysis. Here is a <a href=https://bioconductor.org/packages/release/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf>link</a> to DiffBind's documentation.</td> </tr> </tbody> </table> <p><strong>References</strong></p> <p><sup>1. <strong>MACS:</strong> Zhang, Y., et al. (2008). Model-based Analysis of ChIP-Seq (MACS). Genome Biol 9: R137. <a href=https://github.com/macs3-project/MACS>https://github.com/macs3-project/MACS</a></sup><br> <sup>2. <strong>Sicer:</strong> Xu S., et al. (2014). Spatial Clustering for Identification of ChIP-Enriched Regions (SICER) to Map Regions of Histone Methylation Patterns in Embryonic Stem Cells. Methods Mol Biol 1150: 97--111. <a href=https://zanglab.github.io/SICER2/ >https://zanglab.github.io/SICER2/</a></sup><br> <sup>3. <strong>Genrich:</strong> Gaspar,J.M. (2018) Genrich: Detecting sites of genomic enrichment. <a href=https://github.com/jsh58/Genrich>https://github.com/jsh58/Genrich</a>.</sup><br> <sup>4. <strong>SEACR:</strong> Meers MP, Tenenbaum D, Henikoff S. (2019). Peak calling by Sparse Enrichment Analysis for CUT&amp;RUN chromatin profiling. Epigenetics and Chromatin 12(1):42. <a href=https://github.com/FredHutch/SEACR>https://github.com/FredHutch/SEACR</a></sup><br> <sup>5. <strong>MANorm:</strong> Shao, Z., et al. (2012). MAnorm: a robust model for quantitative comparison of ChIP-Seq data sets. Genome Biology 13: R16. <a href=https://manorm.readthedocs.io/en/latest/index.html>https://manorm.readthedocs.io/en/latest/index.html</a></sup><br> <sup>6. <strong>DiffBind:</strong> Ross-Innes C.S., et al. (2012). Differential oestrogen receptor binding is associated with clinical outcome in breast cancer. Nature 481: 389--393.</sup><br> <sup>7. <strong>DiffBind:</strong> Stark R. and G. Brown. (2011). DiffBind: differential binding analysis of ChIP-Seq peak data. <a href=http://bioconductor.org/packages/release/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf>http://bioconductor.org/packages/release/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf</a></sup><br> <sup>8. <strong>Deseq2:</strong> Love, Michael I et al. “Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2.” Genome biology vol. 15,12 (2014): 550.</sup> </p> <h2 id=annotations-motifs-and-qc-metrics>Annotations, motifs, and QC metrics<a class=headerlink href=#annotations-motifs-and-qc-metrics title="Permanent link">&para;</a></h2> <p>The pipeline includes tools for peak annotation, motif calling, and quality-control metrics. The tools used in the pipeline are listed below, along with their versions and a brief description of their purpose.</p> <table> <thead> <tr> <th><strong>Tool</strong></th> <th align=center><strong>Version</strong></th> <th align=left><strong>Notes</strong></th> </tr> </thead> <tbody> <tr> <td>Uropa<sup>1</sup></td> <td align=center>4.0.3</td> <td align=left>Uropa is utilized for peak annotations, providing comprehensive annotation features. Here is a <a href=https://uropa-manual.readthedocs.io/introduction.html>link</a> to Uropa's documentation. See the glossary for options in this pipeline: <a href=https://openomics.github.io/chrom-seek/faq/glossary/#annotation-options>link</a>.</td> </tr> <tr> <td>Homer<sup>2</sup></td> <td align=center>4.11.1</td> <td align=left>Homer is being used for motif calling</td> </tr> <tr> <td>MEME<sup>3</sup></td> <td align=center>5.5.5</td> <td align=left>Employed for motif analysis, the MEME suite includes MEME-ChIP for <em>de novo</em> motif discovery and AME for known motif analysis. Note that the Centrimo subcomponent of MEME-ChIP may produce inaccurate results for broad peak calling tools. Here is a <a href=https://meme-suite.org/meme/index.html>link</a> to MEME suite's documentation.</td> </tr> <tr> <td>IDR<sup>4</sup></td> <td align=center>2.0.3</td> <td align=left>One method for identifying consensus peaks. Only works for 2 replicates. This tool is not currently in the pipeline.</td> </tr> <tr> <td>Jaccard</td> <td align=center>NA</td> <td align=left>Calculation of peak call consistency between two conditions. This is currently is not included in the pipeline. Requires: <code>pybedtools</code> <code>pysam</code>.</td> </tr> <tr> <td>FRiP</td> <td align=center>NA</td> <td align=left>Represents the <em>"fraction of reads in peaks"</em> calculated as the proportion of aligned reads falling within the peaks called by specific tool for a given sample. Requires: <code>pybedtools</code> <code>pysam</code></td> </tr> </tbody> </table> <p><strong>References</strong></p> <p><sup>1. <strong>Uropa:</strong> Kondili M., et al. (2017). UROPA: a tool for Universal RObust Peak Annotation. Scientific Reports 7: 2593.</sup><br> <sup>2. <strong>Homer:</strong> Heinz S., et al. (2010). Simple combinations of lineage-determining transcription factors prime cis-regulatory elements required for macrophage and B cell identities. Mol Cell 38(4): 576--589.</sup><br> <sup>3. <strong>MEME:</strong> Bailey T.L., et al. (2015). The MEME Suite. Nucleic Acids Res 43(W1):W39-W49.</sup><br> <sup>4. <strong>IDR:</strong> Li Q., et al. (2011). Measuring reproducibility of high-throughput experiments. Ann Appl Stat 5(3): 1752-1779.</sup> </p> <h2 id=reference-genome-resources>Reference genome resources<a class=headerlink href=#reference-genome-resources title="Permanent link">&para;</a></h2> <p><strong>Blacklist citation</strong>: The ENCODE Project Consortium., Moore, J.E., Purcaro, M.J. et al. Expanded encyclopaedias of DNA elements in the human and mouse genomes. Nature 583, 699–710 (2020).</p> <p><strong>excluderanges citation</strong>: Jonathan D Ogata, Wancen Mu, Eric S Davis, Bingjie Xue, J Chuck Harrell, Nathan C Sheffield, Douglas H Phanstiel, Michael I Love, Mikhail G Dozmorov, excluderanges: exclusion sets for T2T-CHM13, GRCm39, and other genome assemblies, Bioinformatics, Volume 39, Issue 4, April 2023, btad198, <a href=https://doi.org/10.1093/bioinformatics/btad198>https://doi.org/10.1093/bioinformatics/btad198</a></p> <table> <thead> <tr> <th><strong>Reference</strong></th> <th align=center><strong>Version</strong></th> <th align=left><strong>Website</strong></th> </tr> </thead> <tbody> <tr> <td>hg19</td> <td align=center>GrCh37.p13</td> <td align=left><a href=https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_19/GRCh37.p13.genome.fa.gz>https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_19/GRCh37.p13.genome.fa.gz</a></td> </tr> <tr> <td></td> <td align=center>Gencode v19</td> <td align=left><a href=https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz>https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz</a></td> </tr> <tr> <td></td> <td align=center>Blacklist v1</td> <td align=left><a href=https://www.encodeproject.org/files/ENCFF356LFX/@@download/ENCFF356LFX.bed.gz>https://www.encodeproject.org/files/ENCFF356LFX/@@download/ENCFF356LFX.bed.gz</a></td> </tr> <tr> <td></td> <td align=center>Enhancer reference</td> <td align=left><a href=https://fantom.gsc.riken.jp/5/datafiles/latest/extra/Enhancers/human_permissive_enhancers_phase_1_and_2.bed.gz>https://fantom.gsc.riken.jp/5/datafiles/latest/extra/Enhancers/human_permissive_enhancers_phase_1_and_2.bed.gz</a></td> </tr> <tr> <td>hg38</td> <td align=center>GRCh38.p12</td> <td align=left><a href=https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_28/GRCh38.p12.genome.fa.gz>https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_28/GRCh38.p12.genome.fa.gz</a></td> </tr> <tr> <td></td> <td align=center>Gencode v28</td> <td align=left><a href=https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_28/gencode.v28.annotation.gtf.gz>https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_28/gencode.v28.annotation.gtf.gz</a></td> </tr> <tr> <td></td> <td align=center>Blacklist v2</td> <td align=left><a href=http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/hg38-human/hg38.blacklist.bed.gz>http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/hg38-human/hg38.blacklist.bed.gz</a></td> </tr> <tr> <td></td> <td align=center>Enhancer reference</td> <td align=left><a href=https://fantom.gsc.riken.jp/5/datafiles/reprocessed/hg38_v9/extra/CAGE_peaks/hg38_fair_CAGE_peaks_combined_phase1+2.bed.gz>https://fantom.gsc.riken.jp/5/datafiles/reprocessed/hg38_v9/extra/CAGE_peaks/hg38_fair_CAGE_peaks_combined_phase1+2.bed.gz</a></td> </tr> <tr> <td>mm10</td> <td align=center>GRCm38.p6</td> <td align=left><a href=https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M18/GRCm38.p6.genome.fa.gz>https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M18/GRCm38.p6.genome.fa.gz</a></td> </tr> <tr> <td></td> <td align=center>Gencode vM18</td> <td align=left><a href=https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M18/gencode.vM18.annotation.gtf.gz>https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M18/gencode.vM18.annotation.gtf.gz</a></td> </tr> <tr> <td></td> <td align=center>Blacklist v2</td> <td align=left><a href=http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz>http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz</a></td> </tr> <tr> <td></td> <td align=center>Enhancer reference</td> <td align=left><a href=https://fantom.gsc.riken.jp/5/datafiles/reprocessed/mm10_v9/extra/CAGE_peaks/mm10_fair+new_CAGE_peaks_phase1and2.bed.gz>https://fantom.gsc.riken.jp/5/datafiles/reprocessed/mm10_v9/extra/CAGE_peaks/mm10_fair+new_CAGE_peaks_phase1and2.bed.gz</a></td> </tr> <tr> <td>rheMac10</td> <td align=center>Ensembl 110</td> <td align=left>Chr1 example: <a href=https://ftp.ensembl.org/pub/release-110/fasta/macaca_mulatta/dna/Macaca_mulatta.Mmul_10.dna.primary_assembly.1.fa.gz>https://ftp.ensembl.org/pub/release-110/fasta/macaca_mulatta/dna/Macaca_mulatta.Mmul_10.dna.primary_assembly.1.fa.gz</a></td> </tr> <tr> <td></td> <td align=center>Ensembl 110</td> <td align=left><a href=https://ftp.ensembl.org/pub/release-110/gtf/macaca_mulatta/Macaca_mulatta.Mmul_10.110.chr.gtf.gz>https://ftp.ensembl.org/pub/release-110/gtf/macaca_mulatta/Macaca_mulatta.Mmul_10.110.chr.gtf.gz</a></td> </tr> <tr> <td>mm39</td> <td align=center>GRCm39</td> <td align=left><a href=https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M37/GRCm39.primary_assembly.genome.fa.gz>https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M37/GRCm39.primary_assembly.genome.fa.gz</a></td> </tr> <tr> <td></td> <td align=center>Gencode vM37</td> <td align=left><a href=https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M37/gencode.vM37.basic.annotation.gtf.gz>https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M37/gencode.vM37.basic.annotation.gtf.gz</a></td> </tr> </tbody> </table> <p>Note: All fastas and gtfs in built into this pipeline only include reference chromosomes, ie autosomes and sex chromosomes.<br> See: <a href=https://github.com/Boyle-Lab/Blacklist/tree/master/lists>https://github.com/Boyle-Lab/Blacklist/tree/master/lists</a> for all v1 and v2 blacklists available.</p> <h2 id=main-differences-between-assays>Main differences between assays<a class=headerlink href=#main-differences-between-assays title="Permanent link">&para;</a></h2> <ul> <li><strong>ChIP</strong></li> <li>Only assay that can handle SE data </li> <li>Peak callers are: macs Narrow and macs Broad </li> <li><strong>cfChIP</strong></li> <li>Peak caller is: macs Narrow </li> <li>Only runs on hg19</li> <li>Filters chromosome X</li> <li>Changes to annotation conditions</li> <li>UMAP plot created in the DiffBindQC report </li> <li>Some rules run in the other versions do not run here</li> <li><strong>ATAC</strong></li> <li>Peak callers are: macs Narrow and Genrich </li> <li><strong>cutnrun</strong> </li> <li>Peak callers are: macs Narrow and SEACR </li> </ul> <h2 id=acknowledgements>Acknowledgements<a class=headerlink href=#acknowledgements title="Permanent link">&para;</a></h2> <h3 id=biowulf>Biowulf<a class=headerlink href=#biowulf title="Permanent link">&para;</a></h3> <p>If you <a href=https://hpc.nih.gov/Research/ >utilized NIH's Biowulf cluster</a> to run <code>chrom-seek</code>, <em>please do not forget to provide an acknowlegement</em>! The continued growth and support of NIH's Biowulf cluster is dependent upon its demonstrable value to the NIH Intramural Research Program. If you publish research that involved significant use of Biowulf, please cite the cluster.</p> <p><strong>Suggested citation text</strong></p> <div class=highlight><pre><span></span><code>This work utilized the computational resources of the NIH HPC Biowulf cluster. (http://hpc.nih.gov)
</code></pre></div> <hr> <div class=md-source-file> <small> Last update: 2025-09-29 </small> </div> </article> </div> </div> <a href=# class="md-top md-top--hidden md-icon" data-md-component=top> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../best_practices/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Best Practices" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Best Practices </div> </div> </a> <a href=../faq/glossary/ class="md-footer__link md-footer__link--next" aria-label="Next: Glossary of terms" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Glossary of terms </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2024 OpenOmics </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://idss-bioinformatics.nih.gov/ target=_blank rel=noopener title=idss-bioinformatics.nih.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M319.9 320c57.41 0 103.1-46.56 103.1-104s-46.54-104-103.1-104c-57.41 0-103.1 46.56-103.1 104-.9 57.4 45.7 104 103.1 104zm50 32h-99.8C191.6 352 128 411.7 128 485.3c0 14.8 12.7 26.7 28.4 26.7h327.2c15.7 0 28.4-11.9 28.4-26.7 0-73.6-63.6-133.3-142.1-133.3zM512 160c44.18 0 80-35.82 80-80S556.2 0 512 0c-44.18 0-80 35.82-80 80s35.8 80 80 80zm-328.1 56c0-5.449.982-10.63 1.609-15.91C174.6 194.1 162.6 192 149.9 192H88.08C39.44 192 0 233.8 0 285.3 0 295.6 7.887 304 17.62 304h199.5c-20.42-23.8-33.22-54.3-33.22-88zM128 160c44.18 0 80-35.82 80-80S172.2 0 128 0C83.82 0 48 35.82 48 80s35.82 80 80 80zm423.9 32h-61.84c-12.8 0-24.88 3.037-35.86 8.24.6 5.26 1.6 10.36 1.6 15.76 0 33.71-12.78 64.21-33.16 88h199.7c9.76 0 17.66-8.4 17.66-18.7 0-51.5-39.4-93.3-88.1-93.3z"/></svg> </a> <a href=https://github.com/OpenOmics target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://hub.docker.com/u/skchronicles target=_blank rel=noopener title=hub.docker.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1v59.4zm0-204.3h-66.1v60.7h66.1V32zm78.2 144.8H362v59.4h66.1v-59.4zm-156.3-72.1h-66.1v60.1h66.1v-60.1zm78.1 0h-66.1v60.1h66.1v-60.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1l-13.3-8.9zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm-78.1-72.1h-66.1v60.1h66.1v-60.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["navigation.tabs", "navigation.top", "toc.integrate"], "search": "../assets/javascripts/workers/search.b028fd86.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../assets/javascripts/bundle.0f659f14.min.js></script> </body> </html>