<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="An awesome OpenOmics chrom-seek pipeline for cfChIP-seq, ChIP-seq, and ATAC-seq data"><meta name=author content="Skyler Kuhn"><link rel=icon href=../assets/favicon/favicon.ico><meta name=generator content="mkdocs-1.4.3, mkdocs-material-8.3.0"><title>Manually running DiffBind - chrom-seek</title><link rel=stylesheet href=../assets/stylesheets/main.fd896c8a.min.css><link rel=stylesheet href=../assets/stylesheets/palette.cbb835fc.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../css/extra.css><script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#chrom-seek-manual-diffbind-analyses-and-downstream-steps class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-component=outdated hidden> <aside class="md-banner md-banner--warning"> </aside> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=.. title=chrom-seek class="md-header__button md-logo" aria-label=chrom-seek data-md-component=logo> <img src=../assets/icons/doc-book.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> chrom-seek </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Manually running DiffBind </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 6a6 6 0 0 1 6 6c0 2.22-1.21 4.16-3 5.2V19a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1v-1.8c-1.79-1.04-3-2.98-3-5.2a6 6 0 0 1 6-6m2 15v1a1 1 0 0 1-1 1h-2a1 1 0 0 1-1-1v-1h4m6-10h3v2h-3v-2M1 11h3v2H1v-2M13 1v3h-2V1h2M4.92 3.5l2.13 2.14-1.42 1.41L3.5 4.93 4.92 3.5m12.03 2.13 2.12-2.13 1.43 1.43-2.13 2.12-1.42-1.42Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3 3.19.09m3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95 2.06.05m-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/OpenOmics/chrom-seek title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> OpenOmics/chrom-seek </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=.. class=md-tabs__link> About </a> </li> <li class=md-tabs__item> <a href=../usage/run/ class=md-tabs__link> Commands </a> </li> <li class=md-tabs__item> <a href=../best_practices/ class=md-tabs__link> Best Practices </a> </li> <li class=md-tabs__item> <a href=../tools_software/ class=md-tabs__link> Tools/References </a> </li> <li class=md-tabs__item> <a href=../faq/glossary/ class="md-tabs__link md-tabs__link--active"> FAQ </a> </li> <li class=md-tabs__item> <a href=../license/ class=md-tabs__link> License </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=.. title=chrom-seek class="md-nav__button md-logo" aria-label=chrom-seek data-md-component=logo> <img src=../assets/icons/doc-book.svg alt=logo> </a> chrom-seek </label> <div class=md-nav__source> <a href=https://github.com/OpenOmics/chrom-seek title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> OpenOmics/chrom-seek </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=.. class=md-nav__link> About </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_2 type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2> Commands <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Commands data-md-level=1> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Commands </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../usage/run/ class=md-nav__link> chrom-seek run </a> </li> <li class=md-nav__item> <a href=../usage/unlock/ class=md-nav__link> chrom-seek unlock </a> </li> <li class=md-nav__item> <a href=../usage/install/ class=md-nav__link> chrom-seek install </a> </li> <li class=md-nav__item> <a href=../usage/cache/ class=md-nav__link> chrom-seek cache </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../best_practices/ class=md-nav__link> Best Practices </a> </li> <li class=md-nav__item> <a href=../tools_software/ class=md-nav__link> Tools/References </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_5 type=checkbox id=__nav_5 checked> <label class=md-nav__link for=__nav_5> FAQ <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=FAQ data-md-level=1> <label class=md-nav__title for=__nav_5> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../faq/glossary/ class=md-nav__link> Glossary of terms </a> </li> <li class=md-nav__item> <a href=../Useful_review_articles/ class=md-nav__link> Useful reviews </a> </li> <li class=md-nav__item> <a href=../faq/Input_vs_IgG/ class=md-nav__link> Input vs IgG </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Manually running DiffBind <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> Manually running DiffBind </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> Overview </a> </li> <li class=md-nav__item> <a href=#prerequisites class=md-nav__link> Prerequisites </a> </li> <li class=md-nav__item> <a href=#stage-1-creating-new-diffbind-csv-files class=md-nav__link> Stage 1: Creating New DiffBind CSV Files </a> <nav class=md-nav aria-label="Stage 1: Creating New DiffBind CSV Files"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#purpose class=md-nav__link> Purpose </a> </li> <li class=md-nav__item> <a href=#steps class=md-nav__link> Steps </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#stage-2-running-diffbind-analysis class=md-nav__link> Stage 2: Running DiffBind Analysis </a> <nav class=md-nav aria-label="Stage 2: Running DiffBind Analysis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#purpose_1 class=md-nav__link> Purpose </a> </li> <li class=md-nav__item> <a href=#input-files-required class=md-nav__link> Input Files Required </a> </li> <li class=md-nav__item> <a href=#prepare-run-conditions class=md-nav__link> Prepare Run Conditions </a> </li> <li class=md-nav__item> <a href=#stage-2a-qc-mode-analysis class=md-nav__link> Stage 2A: QC Mode Analysis </a> </li> <li class=md-nav__item> <a href=#stage-2b-differential-peak-calling-analysis class=md-nav__link> Stage 2B: Differential Peak Calling Analysis </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#stage-3-peak-annotation-with-uropa class=md-nav__link> Stage 3: Peak Annotation with UROPA </a> <nav class=md-nav aria-label="Stage 3: Peak Annotation with UROPA"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#purpose_2 class=md-nav__link> Purpose </a> </li> <li class=md-nav__item> <a href=#input-files-required_1 class=md-nav__link> Input Files Required </a> </li> <li class=md-nav__item> <a href=#steps_1 class=md-nav__link> Steps </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#stage-4-merging-annotations-with-differential-results class=md-nav__link> Stage 4: Merging Annotations with Differential Results </a> <nav class=md-nav aria-label="Stage 4: Merging Annotations with Differential Results"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#purpose_3 class=md-nav__link> Purpose </a> </li> <li class=md-nav__item> <a href=#input-files-required_2 class=md-nav__link> Input Files Required </a> </li> <li class=md-nav__item> <a href=#steps_2 class=md-nav__link> Steps </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#getting-help class=md-nav__link> Getting Help </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../faq/questions/ class=md-nav__link> General Questions </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../license/ class=md-nav__link> License </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/OpenOmics/chrom-seek/edit/main/docs/manual_workflows.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1 id=chrom-seek-manual-diffbind-analyses-and-downstream-steps>Chrom-seek Manual: DiffBind Analyses and Downstream Steps<a class=headerlink href=#chrom-seek-manual-diffbind-analyses-and-downstream-steps title="Permanent link">&para;</a></h1> <p>This guide provides step-by-step instructions for performing post-pipeline analysis workflows outside of Snakemake, using the same working directory and scripts available in the <code>/bin</code> directory of chrom-seek.</p> <h2 id=overview>Overview<a class=headerlink href=#overview title="Permanent link">&para;</a></h2> <p>After running the chrom-seek pipeline, you may want to perform additional analysis steps manually, especially if you do not wish to retain the raw FASTQs and sorted BAMs. This guide covers four main stages:</p> <ol> <li><a href=#stage-1-creating-new-diffbind-csv-files>Creating new DiffBind CSV files</a> </li> <li><a href=#stage-2-running-diffbind-analysis>Running DiffBind in QC or differential peak calling mode</a> </li> <li><a href=#stage-3-peak-annotation-with-uropa>Annotating peaks with UROPA</a> </li> <li><a href=#stage-4-merging-annotations-with-differential-results>Merging annotations with differential peak calls</a> </li> </ol> <h2 id=prerequisites>Prerequisites<a class=headerlink href=#prerequisites title="Permanent link">&para;</a></h2> <ul> <li>Completed chrom-seek pipeline run </li> <li>Access to the pipeline working directory </li> <li>Access to the Docker/Singularity image containing DiffBind v2 </li> <li>Python environment (for merging results) </li> </ul> <hr> <h2 id=stage-1-creating-new-diffbind-csv-files>Stage 1: Creating New DiffBind CSV Files<a class=headerlink href=#stage-1-creating-new-diffbind-csv-files title="Permanent link">&para;</a></h2> <h3 id=purpose>Purpose<a class=headerlink href=#purpose title="Permanent link">&para;</a></h3> <p>DiffBind requires a CSV file with metadata (in a specific format), including the locations of Q5DD BAMs and peak files. Manually consolidating this information can be time-consuming and error-prone, but the chrom-seek DiffBindQC rules generate CSVs with all sample information in one place.</p> <h3 id=steps>Steps<a class=headerlink href=#steps title="Permanent link">&para;</a></h3> <ol> <li>Retrieve the relevant CSV from <code>&lt;PIPELINE_OUTPUT_DIR&gt;/PeakQC/DB_QC</code>. </li> <li>Remove rows for samples that should not be included in the analysis. </li> <li>Update <strong>only</strong> the following three columns: </li> <li><strong>CONDITION:</strong> Group information used by DiffBind to define contrasts. <ul> <li><em>Note:</em> Do not start values with numbers. </li> </ul> </li> <li><strong>TREATMENT:</strong> (Optional) Used for blocking/paired analysis when available. <ul> <li><em>Note:</em> Do not start values with numbers. </li> </ul> </li> <li><strong>REPLICATE:</strong> A simple count (1, 2, 3, …) per sample. </li> <li>For <a href=#stage-2-running-diffbind-analysis>differential analysis</a>, sort rows so that the first group (<code>group1</code>) appears before the second (<code>group2</code>) for contrasts (<code>group1-group2</code>). </li> </ol> <p><a href=#overview>⬆ Back to top</a></p> <hr> <h2 id=stage-2-running-diffbind-analysis>Stage 2: Running DiffBind Analysis<a class=headerlink href=#stage-2-running-diffbind-analysis title="Permanent link">&para;</a></h2> <h3 id=purpose_1>Purpose<a class=headerlink href=#purpose_1 title="Permanent link">&para;</a></h3> <p>Use the CSV created in <a href=#stage-1-creating-new-diffbind-csv-files>Stage 1</a> and DiffBind to identify consensus peaks and normalize read counts. </p> <ul> <li><strong>QC mode analysis</strong> produces a TMM-normalized counts table, suitable for use with external differential peak callers or for generating heatmaps. It also calculates PCA and UMAP coordinates. </li> <li><strong>Differential peak calling analysis</strong> uses DiffBind to identify differential peaks, as done in the chrom-seek pipeline. </li> </ul> <h3 id=input-files-required>Input Files Required<a class=headerlink href=#input-files-required title="Permanent link">&para;</a></h3> <ul> <li>DiffBind CSV file (from <a href=#stage-1-creating-new-diffbind-csv-files>Stage 1</a>) </li> <li>BAM (and associated BAI) files and peak files referenced in the CSV </li> </ul> <h3 id=prepare-run-conditions>Prepare Run Conditions<a class=headerlink href=#prepare-run-conditions title="Permanent link">&para;</a></h3> <ol> <li> <p>Request an interactive session (adjust resources if needed):<br> <div class=highlight><pre><span></span><code>sinteractive<span class=w> </span>-N<span class=w> </span><span class=m>1</span><span class=w> </span>-n<span class=w> </span><span class=m>1</span><span class=w> </span>--time<span class=o>=</span><span class=m>1</span>-12:00:00<span class=w> </span>--mem<span class=o>=</span>100G<span class=w> </span>--gres<span class=o>=</span>lscratch:200<span class=w> </span>--cpus-per-task<span class=o>=</span><span class=m>4</span>
</code></pre></div></p> </li> <li> <p>Initialize the Singularity container for DiffBind v2. </p> </li> </ol> <blockquote> <p><strong>Note</strong><br> These conditions ensure lscratch remains accessible and that the Stage 1 CSV file can retain full paths without causing errors. </p> </blockquote> <p><strong>Usage</strong>:<br> <div class=highlight><pre><span></span><code>singularity<span class=w> </span>run<span class=w> </span><span class=se>\</span>
<span class=w>  </span>-C<span class=w> </span><span class=se>\</span>
<span class=w>  </span>-e<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--env<span class=w> </span><span class=nv>TMPDIR</span><span class=o>=</span>/tmp,TMP<span class=o>=</span>/tmp<span class=w> </span><span class=se>\</span>
<span class=w>  </span>-B<span class=w> </span>/lscratch/<span class=nv>$SLURM_JOBID</span>:/tmp,&lt;PROJECT_WORKING_DIR&gt;:&lt;PROJECT_WORKING_DIR&gt;:rw<span class=w> </span><span class=se>\</span>
<span class=w>  </span>docker://skchronicles/cfchip_toolkit:v0.5.0<span class=w> </span><span class=se>\</span>
<span class=w>  </span>bash
</code></pre></div></p> <p><strong>Example</strong>:<br> <div class=highlight><pre><span></span><code><span class=nv>working_dir</span><span class=o>=</span>/data/OpenOmics/dev/datasets/outputs/test_homer
singularity<span class=w> </span>run<span class=w> </span><span class=se>\</span>
<span class=w>  </span>-C<span class=w> </span><span class=se>\</span>
<span class=w>  </span>-e<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--env<span class=w> </span><span class=nv>TMPDIR</span><span class=o>=</span>/tmp,TMP<span class=o>=</span>/tmp,working_dir<span class=o>=</span><span class=nv>$working_dir</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>-B<span class=w> </span>/lscratch/<span class=nv>$SLURM_JOBID</span>:/tmp,<span class=nv>$working_dir</span>:<span class=nv>$working_dir</span>:rw<span class=w> </span><span class=se>\</span>
<span class=w>  </span>docker://skchronicles/cfchip_toolkit:v0.5.0<span class=w> </span><span class=se>\</span>
<span class=w>  </span>bash
</code></pre></div></p> <h3 id=stage-2a-qc-mode-analysis>Stage 2A: QC Mode Analysis<a class=headerlink href=#stage-2a-qc-mode-analysis title="Permanent link">&para;</a></h3> <ol> <li>Run <code>DiffBind_v2_QC.Rmd</code>. </li> </ol> <blockquote> <p><strong>Notes</strong><br> - Rmarkdown requires full paths; otherwise it looks for <code>&lt;INPUT_CSV_FILE&gt;</code> in <code>&lt;WORKING_DIR&gt;/bin</code>.<br> - The <code>&lt;PEAK_TOOL&gt;</code> token must be one of: <code>macsNarrow</code>, <code>Genrich</code>, <code>macsBroad</code>, <code>SEACR</code>. </p> </blockquote> <p><strong>Usage</strong>:<br> <div class=highlight><pre><span></span><code>Rscript<span class=w> </span>-e<span class=w> </span><span class=s1>&#39;rmarkdown::render(&quot;&lt;WORKING_DIR&gt;/bin/DiffBind_v2_QC.Rmd&quot;, </span>
<span class=s1>            output_file=&quot;&lt;OUTPUT_HTML_FILE&gt;&quot;,</span>
<span class=s1>            params=list(csvfile=&quot;&lt;INPUT_CSV_FILE&gt;&quot;, counts_bed=&quot;&lt;OUTPUT_BED_FILE&gt;&quot;, </span>
<span class=s1>            counts_csv=&quot;&lt;OUTPUT_COUNTS_CSV_FILE&gt;&quot;, peakcaller=&quot;&lt;PEAK_TOOL&gt;&quot;))&#39;</span>
</code></pre></div></p> <p><strong>Example</strong>:<br> <div class=highlight><pre><span></span><code>Rscript<span class=w> </span>-e<span class=w> </span><span class=s1>&#39;rmarkdown::render(&quot;/data/OpenOmics/dev/datasets/outputs/test_homer/bin/DiffBind_v2_QC.Rmd&quot;, </span>
<span class=s1>  output_file=&quot;/data/OpenOmics/dev/datasets/outputs/test_homer/PeakQC/DB_QC/AllSamples-macsNarrow/AllSamples-macsNarrow_DiffBindQC.html&quot;,</span>
<span class=s1>  params=list(csvfile=&quot;/data/OpenOmics/dev/datasets/outputs/test_homer/PeakQC/DB_QC/AllSamples-macsNarrow/AllSamples-macsNarrow_DiffBind_prep.csv&quot;, </span>
<span class=s1>  counts_bed=&quot;/data/OpenOmics/dev/datasets/outputs/test_homer/PeakQC/DB_QC/AllSamples-macsNarrow/AllSamples-macsNarrow_DiffBindQC_TMMcounts.bed&quot;, </span>
<span class=s1>  counts_csv=&quot;/data/OpenOmics/dev/datasets/outputs/test_homer/PeakQC/DB_QC/AllSamples-macsNarrow/AllSamples-macsNarrow_DiffBindQC_TMMcounts.csv&quot;, </span>
<span class=s1>  peakcaller=&quot;macsNarrow&quot;))&#39;</span>
</code></pre></div></p> <h3 id=stage-2b-differential-peak-calling-analysis>Stage 2B: Differential Peak Calling Analysis<a class=headerlink href=#stage-2b-differential-peak-calling-analysis title="Permanent link">&para;</a></h3> <ol> <li>Load counts using <code>DiffBind_v2_load.R</code>. </li> </ol> <blockquote> <p><strong>Note</strong><br> The <code>&lt;PEAK_TOOL&gt;</code> token must be one of: <code>macsNarrow</code>, <code>Genrich</code>, <code>macsBroad</code>, <code>SEACR</code>.</p> </blockquote> <p><strong>Usage</strong>:<br> <div class=highlight><pre><span></span><code>&lt;WORKING_DIR&gt;/bin/DiffBind_v2_load.R<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--csvfile<span class=w> </span>&lt;INPUT_CSV_FILE&gt;<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--counts<span class=w> </span>&lt;INPUT_RDS_FILE&gt;<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--list<span class=w> </span>&lt;OUTPUT_PEAK_BED_FILE&gt;<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--peakcaller<span class=w> </span>&lt;PEAK_TOOL&gt;
</code></pre></div></p> <p><strong>Example</strong>:<br> <div class=highlight><pre><span></span><code>/data/OpenOmics/dev/datasets/outputs/test_homer/bin/DiffBind_v2_load.R<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--csvfile<span class=w> </span>/data/OpenOmics/dev/datasets/outputs/test_homer/DiffBind/IFN0h_vs_IFN24h-macsBroad/IFN0h_vs_IFN24h-macsBroad_Diffbind_prep.csv<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--counts<span class=w> </span>/data/OpenOmics/dev/datasets/outputs/test_homer/DiffBind/IFN0h_vs_IFN24h-macsBroad/IFN0h_vs_IFN24h-macsBroad_Diffbind_counts.rds<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--list<span class=w> </span>/data/OpenOmics/dev/datasets/outputs/test_homer/DiffBind/IFN0h_vs_IFN24h-macsBroad/IFN0h_vs_IFN24h-macsBroad_Diffbind_fullList.bed<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--peakcaller<span class=w> </span>macsBroad
</code></pre></div></p> <ol> <li>Execute differential comparisons using <strong>DiffBind v2.15.2</strong>.</li> </ol> <p>1) Choose the method: <strong>DESeq2</strong> or <strong>edgeR</strong>. DESeq2 generally suggested. Consider EdgeR when there is large variance in library size, library size is confounding, or there are prior knowledge suggesting that there should be an equivalent number of peaks on both sides of the contrast.<br> 2) Decide whether your design requires <strong>no blocking</strong> or <strong>blocking</strong> (paired or batch effects).<br> 3) Select the relevant script (<code>&lt;ANALYSIS_SCRIPT&gt;</code>):<br> - <code>&lt;WORKING_DIR&gt;/bin/DiffBind_v2_Deseq2.Rmd</code><br> - <code>&lt;WORKING_DIR&gt;/bin/DiffBind_v2_Deseq2_block.Rmd</code><br> - <code>&lt;WORKING_DIR&gt;/bin/DiffBind_v2_EdgeR.Rmd</code><br> - <code>&lt;WORKING_DIR&gt;/bin/DiffBind_v2_EdgeR_block.Rmd</code><br> 4) Define the group contrast from your experimental setup: <code>"{group1}_vs_{group2}"</code> → <code>&lt;INPUT_CONTRASTS&gt;</code>.<br> 5) Render the report: </p> <div class=highlight><pre><span></span><code>  &gt; **Note**  
  &gt; The `&lt;PEAK_TOOL&gt;` token must be one of: `macsNarrow`, `Genrich`, `macsBroad`, `SEACR`.

  **Usage**:  
  ```bash
  Rscript -e &#39;rmarkdown::render(&quot;&lt;ANALYSIS_SCRIPT&gt;&quot;, 
              output_file=&quot;&lt;OUTPUT_DIFFBIND_REPORT_FILE&gt;&quot;,
              params=list(csvfile=&quot;&lt;INPUT_CSV_FILE&gt;&quot;, 
              peakcaller=&quot;&lt;PEAK_TOOL&gt;&quot;, 
              list_file=&quot;&lt;OUTPUT_PEAK_BED_LIST&gt;&quot;,
              up_file=&quot;&lt;OUTPUT_UP_REGULATED_FILE&gt;&quot;, 
              down_file=&quot;&lt;OUTPUT_DOWN_REGULATED_FILE&gt;&quot;, 
              contrasts=&quot;&lt;INPUT_CONTRASTS&gt;&quot;, 
              counts=&quot;&lt;INPUT_PEAK_COUNTS_RDS_FILE&gt;&quot;))&#39;
  ```

  **Example**:  
  ```bash
  Rscript -e &#39;rmarkdown::render(&quot;/data/OpenOmics/dev/datasets/outputs/test_homer/bin/DiffBind_v2_Deseq2.Rmd&quot;, 
    output_file=&quot;/data/OpenOmics/dev/datasets/outputs/test_homer/DiffBind/IFN0h_vs_IFN24h-macsNarrow/IFN0h_vs_IFN24h-macsNarrow_Diffbind_DeSeq2.html&quot;, 
    params=list(csvfile=&quot;/data/OpenOmics/dev/datasets/outputs/test_homer/DiffBind/IFN0h_vs_IFN24h-macsNarrow/IFN0h_vs_IFN24h-macsNarrow_Diffbind_prep.csv&quot;, 
    peakcaller=&quot;macsNarrow&quot;, 
    list_file=&quot;/data/OpenOmics/dev/datasets/outputs/test_homer/DiffBind/IFN0h_vs_IFN24h-macsNarrow/IFN0h_vs_IFN24h-macsNarrow_Diffbind_DeSeq2_peak_list.tab&quot;, 
    up_file=&quot;/data/OpenOmics/dev/datasets/outputs/test_homer/DiffBind/IFN0h_vs_IFN24h-macsNarrow/IFN0h_vs_IFN24h-macsNarrow_Diffbind_DeSeq2_up.bed&quot;, 
    down_file=&quot;/data/OpenOmics/dev/datasets/outputs/test_homer/DiffBind/IFN0h_vs_IFN24h-macsNarrow/IFN0h_vs_IFN24h-macsNarrow_Diffbind_DeSeq2_down.bed&quot;, 
    contrasts=&quot;IFN0h_vs_IFN24h&quot;, 
    counts=&quot;/data/OpenOmics/dev/datasets/outputs/test_homer/DiffBind/IFN0h_vs_IFN24h-macsNarrow/IFN0h_vs_IFN24h-macsNarrow_Diffbind_counts.rds&quot;))&#39;
  ```
</code></pre></div> <p><a href=#overview>⬆ Back to top</a></p> <hr> <h2 id=stage-3-peak-annotation-with-uropa>Stage 3: Peak Annotation with UROPA<a class=headerlink href=#stage-3-peak-annotation-with-uropa title="Permanent link">&para;</a></h2> <h3 id=purpose_2>Purpose<a class=headerlink href=#purpose_2 title="Permanent link">&para;</a></h3> <p>Annotate genomic peaks with gene information, regulatory elements, and genomic context using UROPA.</p> <h3 id=input-files-required_1>Input Files Required<a class=headerlink href=#input-files-required_1 title="Permanent link">&para;</a></h3> <ul> <li>Peak file (BED or narrowPeak format) <code>&lt;PEAK_FILE&gt;</code> </li> <li>GTF annotation file <code>&lt;GTF_FILE&gt;</code> </li> <li>UROPA configuration file <code>&lt;UROPA_CONFIG&gt;</code> </li> </ul> <p>The GTF file used in the pipeline can be found in <code>config.json</code> under <code>[references][&lt;SPECIES&gt;][GTFFILE]</code>.</p> <h3 id=steps_1>Steps<a class=headerlink href=#steps_1 title="Permanent link">&para;</a></h3> <ol> <li>Create a UROPA configuration file. The example below is what is currently used for chrom-seek when assay is not <code>cfchip</code>.</li> </ol> <div class=highlight><pre><span></span><code><span class=p>{</span>
<span class=w>  </span><span class=nt>&quot;queries&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span>
<span class=w>    </span><span class=p>{</span>
<span class=w>      </span><span class=nt>&quot;feature&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span><span class=s2>&quot;gene&quot;</span><span class=p>],</span>
<span class=w>      </span><span class=nt>&quot;filter_attribute&quot;</span><span class=p>:</span><span class=w> </span><span class=s2>&quot;gene_type&quot;</span><span class=p>,</span>
<span class=w>      </span><span class=nt>&quot;attribute_values&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span><span class=s2>&quot;protein_coding&quot;</span><span class=p>],</span>
<span class=w>      </span><span class=nt>&quot;feature_anchor&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span><span class=s2>&quot;start&quot;</span><span class=p>],</span>
<span class=w>      </span><span class=nt>&quot;distance&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span><span class=mi>3000</span><span class=p>,</span><span class=w> </span><span class=mi>1000</span><span class=p>],</span>
<span class=w>      </span><span class=nt>&quot;name&quot;</span><span class=p>:</span><span class=w> </span><span class=s2>&quot;query_1&quot;</span>
<span class=w>    </span><span class=p>},</span>
<span class=w>    </span><span class=p>{</span>
<span class=w>      </span><span class=nt>&quot;feature&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span><span class=s2>&quot;gene&quot;</span><span class=p>],</span>
<span class=w>      </span><span class=nt>&quot;filter_attribute&quot;</span><span class=p>:</span><span class=w> </span><span class=s2>&quot;gene_type&quot;</span><span class=p>,</span>
<span class=w>      </span><span class=nt>&quot;attribute_values&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span><span class=s2>&quot;protein_coding&quot;</span><span class=p>],</span>
<span class=w>      </span><span class=nt>&quot;feature_anchor&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span><span class=s2>&quot;start&quot;</span><span class=p>],</span>
<span class=w>      </span><span class=nt>&quot;distance&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span><span class=mi>10000</span><span class=p>],</span>
<span class=w>      </span><span class=nt>&quot;name&quot;</span><span class=p>:</span><span class=w> </span><span class=s2>&quot;query_2&quot;</span>
<span class=w>    </span><span class=p>},</span>
<span class=w>    </span><span class=p>{</span>
<span class=w>      </span><span class=nt>&quot;feature&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span><span class=s2>&quot;gene&quot;</span><span class=p>],</span>
<span class=w>      </span><span class=nt>&quot;filter_attribute&quot;</span><span class=p>:</span><span class=w> </span><span class=s2>&quot;gene_type&quot;</span><span class=p>,</span>
<span class=w>      </span><span class=nt>&quot;attribute_values&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span><span class=s2>&quot;protein_coding&quot;</span><span class=p>],</span>
<span class=w>      </span><span class=nt>&quot;feature_anchor&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span><span class=s2>&quot;start&quot;</span><span class=p>],</span>
<span class=w>      </span><span class=nt>&quot;distance&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span><span class=mi>100000</span><span class=p>],</span>
<span class=w>      </span><span class=nt>&quot;name&quot;</span><span class=p>:</span><span class=w> </span><span class=s2>&quot;query_3&quot;</span>
<span class=w>    </span><span class=p>}</span>
<span class=w>  </span><span class=p>],</span>
<span class=w>  </span><span class=nt>&quot;show_attributes&quot;</span><span class=p>:</span><span class=w> </span><span class=p>[</span>
<span class=w>    </span><span class=s2>&quot;gene_id&quot;</span><span class=p>,</span>
<span class=w>    </span><span class=s2>&quot;gene_name&quot;</span><span class=p>,</span>
<span class=w>    </span><span class=s2>&quot;gene_type&quot;</span>
<span class=w>  </span><span class=p>],</span>
<span class=w>  </span><span class=nt>&quot;priority&quot;</span><span class=p>:</span><span class=w> </span><span class=s2>&quot;True&quot;</span><span class=p>,</span>
<span class=w>  </span><span class=nt>&quot;gtf&quot;</span><span class=p>:</span><span class=w> </span><span class=s2>&quot;&lt;GTF_FILE&gt;&quot;</span><span class=p>,</span>
<span class=w>  </span><span class=nt>&quot;bed&quot;</span><span class=p>:</span><span class=w> </span><span class=s2>&quot;&lt;PEAK_FILE&gt;&quot;</span>
<span class=p>}</span>
</code></pre></div> <ol> <li>Run UROPA on Biowulf (or in your container). This example uses the <code>uropa</code> module. We suggest 4 threads and 10G memory.</li> </ol> <p><strong>Usage</strong>: <div class=highlight><pre><span></span><code>module<span class=w> </span>load<span class=w> </span>uropa/4.0.2
uropa<span class=w> </span>-i<span class=w> </span>&lt;UROPA_CONFIG&gt;<span class=w> </span>-t<span class=w> </span><span class=m>4</span>
</code></pre></div></p> <p><strong>Example</strong>: <div class=highlight><pre><span></span><code>module<span class=w> </span>load<span class=w> </span>uropa/4.0.2
uropa<span class=w> </span>-i<span class=w> </span>sample_config.json<span class=w> </span>-t<span class=w> </span><span class=m>4</span>
</code></pre></div></p> <p><a href=#overview>⬆ Back to top</a></p> <hr> <h2 id=stage-4-merging-annotations-with-differential-results>Stage 4: Merging Annotations with Differential Results<a class=headerlink href=#stage-4-merging-annotations-with-differential-results title="Permanent link">&para;</a></h2> <h3 id=purpose_3>Purpose<a class=headerlink href=#purpose_3 title="Permanent link">&para;</a></h3> <p>Combine differential binding results from <a href=#stage-2-running-diffbind-analysis>Stage 2</a> with gene annotations from <a href=#stage-3-peak-annotation-with-uropa>Stage 3</a> to generate comprehensive results tables.</p> <h3 id=input-files-required_2>Input Files Required<a class=headerlink href=#input-files-required_2 title="Permanent link">&para;</a></h3> <ul> <li>DiffBind peak list (<code>&lt;OUTPUT_PEAK_BED_LIST&gt;</code>) from <a href=#stage-2-running-diffbind-analysis>Stage 2</a> </li> <li>UROPA results (<code>&lt;UROPA_FINAL_HITS&gt;</code>) from <a href=#stage-3-peak-annotation-with-uropa>Stage 3</a> </li> </ul> <h3 id=steps_2>Steps<a class=headerlink href=#steps_2 title="Permanent link">&para;</a></h3> <ol> <li>Merge differential results with annotations. Optionally filter at a stricter FDR or fold-change threshold than defaults. Set FDR to 1 for all peaks analyzed.</li> </ol> <p><strong>Usage</strong>: <div class=highlight><pre><span></span><code>python<span class=w> </span>&lt;WORKING_DIR&gt;/bin/merge_diffbind_uropa.py<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--diffbind<span class=w> </span>&lt;DIFFBIND_PEAK_LIST_FILE&gt;<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--uropa<span class=w> </span>&lt;UROPA_FINAL_HITS&gt;<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--fdr<span class=w> </span>&lt;FDR_THRESHOLD&gt;<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--fold<span class=w> </span>&lt;FC_THRESHOLD&gt;<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--output<span class=w> </span>&lt;MERGE_OUTPUT&gt;
</code></pre></div></p> <p><strong>Example</strong>: <div class=highlight><pre><span></span><code>python<span class=w> </span>/data/OpenOmics/dev/datasets/outputs/test_homer/bin/merge_diffbind_uropa.py<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--uropa<span class=w> </span>/data/OpenOmics/dev/datasets/outputs/test_homer/UROPA_annotations/DiffBind/IFN0h_vs_IFN24h-macsNarrow-EdgeR/IFN0h_vs_IFN24h_macsNarrow_EdgeR_protTSS_uropa_finalhits.txt<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--diffbind<span class=w> </span>/data/OpenOmics/dev/datasets/outputs/test_homer/DiffBind/IFN0h_vs_IFN24h-macsNarrow/IFN0h_vs_IFN24h-macsNarrow_Diffbind_EdgeR_peak_list.tab<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--output<span class=w> </span>/data/OpenOmics/dev/datasets/outputs/test_homer/UROPA_DIFFBIND_TBLS/IFN0h_vs_IFN24h-macsNarrow-EdgeR_protTSS_UROPA_DIFFBIND_JOIN.txt<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--fdr<span class=w> </span><span class=m>0</span>.05<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--fold<span class=w> </span><span class=m>0</span>
</code></pre></div></p> <p><a href=#overview>⬆ Back to top</a></p> <hr> <h2 id=getting-help>Getting Help<a class=headerlink href=#getting-help title="Permanent link">&para;</a></h2> <ul> <li><a href=https://hpc.nih.gov/docs/userguide.html>Biowulf user guide</a> (for <code>sinteractive</code>) </li> <li><a href=https://www.rdocumentation.org/packages/DiffBind/versions/2.0.2>DiffBind v2.0.2 documentation</a> </li> <li><a href=https://www.bioconductor.org/packages//2.10/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf>DiffBind v1.2.4 vignette (archival)</a> </li> <li><a href=https://uropa-manual.readthedocs.io/ >UROPA manual</a> </li> </ul> <hr> <div class=md-source-file> <small> Last update: 2025-09-29 </small> </div> </article> </div> </div> <a href=# class="md-top md-top--hidden md-icon" data-md-component=top> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../faq/Input_vs_IgG/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Input vs IgG" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Input vs IgG </div> </div> </a> <a href=../faq/questions/ class="md-footer__link md-footer__link--next" aria-label="Next: General Questions" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> General Questions </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2024 OpenOmics </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://idss-bioinformatics.nih.gov/ target=_blank rel=noopener title=idss-bioinformatics.nih.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M319.9 320c57.41 0 103.1-46.56 103.1-104s-46.54-104-103.1-104c-57.41 0-103.1 46.56-103.1 104-.9 57.4 45.7 104 103.1 104zm50 32h-99.8C191.6 352 128 411.7 128 485.3c0 14.8 12.7 26.7 28.4 26.7h327.2c15.7 0 28.4-11.9 28.4-26.7 0-73.6-63.6-133.3-142.1-133.3zM512 160c44.18 0 80-35.82 80-80S556.2 0 512 0c-44.18 0-80 35.82-80 80s35.8 80 80 80zm-328.1 56c0-5.449.982-10.63 1.609-15.91C174.6 194.1 162.6 192 149.9 192H88.08C39.44 192 0 233.8 0 285.3 0 295.6 7.887 304 17.62 304h199.5c-20.42-23.8-33.22-54.3-33.22-88zM128 160c44.18 0 80-35.82 80-80S172.2 0 128 0C83.82 0 48 35.82 48 80s35.82 80 80 80zm423.9 32h-61.84c-12.8 0-24.88 3.037-35.86 8.24.6 5.26 1.6 10.36 1.6 15.76 0 33.71-12.78 64.21-33.16 88h199.7c9.76 0 17.66-8.4 17.66-18.7 0-51.5-39.4-93.3-88.1-93.3z"/></svg> </a> <a href=https://github.com/OpenOmics target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://hub.docker.com/u/skchronicles target=_blank rel=noopener title=hub.docker.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1v59.4zm0-204.3h-66.1v60.7h66.1V32zm78.2 144.8H362v59.4h66.1v-59.4zm-156.3-72.1h-66.1v60.1h66.1v-60.1zm78.1 0h-66.1v60.1h66.1v-60.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1l-13.3-8.9zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm-78.1-72.1h-66.1v60.1h66.1v-60.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["navigation.tabs", "navigation.top", "toc.integrate"], "search": "../assets/javascripts/workers/search.b028fd86.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../assets/javascripts/bundle.0f659f14.min.js></script> </body> </html>